### DISCUSSION TITLE
Cannot find module '/app/server.js'

### DISCUSSION THREAD

**Discussion by Shivamy2** [adonisjs/core#3309] [Help]
Not able to host on Heroku

```
node server.js
internal/modules/cjs/loader.js:905
throw err;
Error: Cannot find module '/app/server.js'
at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
 at Function.Module._load (internal/modules/cjs/loader.js:746:27)
at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
at internal/main/run_main_module.js:17:47 {
code: 'MODULE_NOT_FOUND',
requireStack: []
2021-11-04T06:04:22.376756+00:00 app[web.1]: }
npm ERR! code ELIFECYCLE
 npm ERR! errno 1
npm ERR! practica-server@1.0.0 start: `node server.js`
 npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the practica-server@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

package.json

```
"scripts": {
    "dev": "node ace serve --watch",
    "build": "node ace build --production",
    "start": "node server.js"
  },
```

**Comment by RomainLanz** [comment]
Hey @Shivamy2! 👋🏻 

We will definitively more information to be able to help you.
In the meanwhile, check [our documentation for Heroku deployment](https://docs.adonisjs.com/cookbooks/deploy-to-heroku).

**Comment by akshaj65** [comment]
i am also getting the same error
if the error was resolved please help me also 
thankyou

**Comment by yevheniib** [comment]
same here, i was getting **_Cannot find module '/app/server.js_**', followed stackoverflow article and added a procfile where specified the path to the file **_web: node src/server.js_**, so now i get *Cannot find module '/app/src/server.js'*, why the **_/app_**, how to specify that that the main file is not in the /app?

**Reply by Gokul-C** [reply]
web: node src/server.js

**Reply by yevheniib** [reply]
i don't get it.
As i said, i've created a Procfile with **_web: node src/server.js_** and now i get the **_Cannot find module '/app/src/server.js'_**. The Procfile works, cause it's picking up the 'src' part from it, but still puts '/app' in the front

**Reply by Gokul-C** [reply]
Create a Procfile with following content

web: node src/server.js 

The Procfile is always a simple text file that is named Procfile exactly. For example, Procfile.txt is not valid.

The Procfile must live in your app’s root directory. It does not function if placed anywhere else.

**Reply by Gokul-C** [reply]
> i don't get it.
> As i said, i've created a Procfile with **_web: node src/server.js_** and now i get the **_Cannot find module '/app/src/server.js'_**. The Procfile works, cause it's picking up the 'src' part from it, but still puts '/app' in the front

app should be in src folder

**Reply by yevheniib** [reply]
sounds about right,

I have the Procfile with no extension after and **_web: node src/server.js_** in it
i have the app.js in the src folder
i have the server.js in the src folder

the only problem i see, might be with that i am using TS, so in there is a "main": "dist/src/server.js" in the package.json and i do the "postinstall" script to compile the files

**Reply by JasminRdev** [reply]
did u resolved this problem? have the same issue & searched all day in internet and couldn't find a solution.

**Reply by yevheniib** [reply]
the problem is gone, but I cannot provide the steps, sorry, I've reset up my express from up to bottom.

**Reply by JasminRdev** [reply]
How is express changing anything when u can't even reach the folder it was supposed to be inside??

**Reply by hansderly** [reply]
@JasminRdev i think the problem was because in the Procfile @yevheniib wrote web: node server.js, but that should be in the script of package.json. The Procfile just need to know the script name and should look like web: npm start, where start is the name of the script in package.json wich run node server.js.
RECAP>>>
package.json:
```
{
...
       "scripts": {
		"start": "node server.js",
		"dev": "nodemon server.js"
	},
...
}
```

Procfile:
web: npm start


Hope that helps.

**Reply by JasminRdev** [reply]
That's for that idea but that procfile wasn't the idea. Something completely different wasn't on a relative path. My Axios baseurl was instead as fixed domain. 

F.e. baseURL = *domain*/API/notes is wrong but baseURL= API/notes was working.

**Reply by JasminRdev** [reply]
Thanks*

**Comment by radoncreep** [comment]
if you're using YARN as your pkg manager and typescript or even js, then;

In my case, I wrote a stupid instruction to use npm instead of yarn in my start script. Its possible npm couldn't read from the yarn.lock file because I included a package I never installed in my project, npm couldn't depict that and the logs from Heroku (sorry can't reproduce) looked like the image from @Shivamy2. 

I changed to "yarn build && node <build directory>" and from the logs on Heroku, yarn was able to depict I didn't install a pkg.
So yea, installing that pkg and deploying worked fine.

_in Procfile_
`web: yarn start`

_in package.json_
`
{
"main": "<outDir value>/server.js", // indicate the entry file for your build, you might call your outputDir "build" or "dist" or any from tsconfig file

"scripts": {
    "start": "yarn run build && node dist/server.js",
    "build": "rimraf ./dist && tsc", // use rimraf to efficiently remove old tsc build
    "dev": "nodemon ./src/server.ts"
  },
`

Hope this helps!

**Comment by observadorxpl** [comment]
Any solution?

**Reply by RomainLanz** [reply]
If you have this error, you are probably trying to run a file that does not exist. As said in the first comment, it is hard for us to help anyone without seeing the proper file structure deployed.

Double check where your `server.js` file resides.

**Comment by observadorxpl** [comment]
I found the problem, just add express dependency haha
![image](https://github.com/user-attachments/assets/50238b62-adf9-4429-8aa7-407f9bab2ec6)
